src/double.jl

```diff
@@ -22,230 +31,234 @@ end
     s = x - y
     v = s - x
     Double(s, (x - (s - v)) + (-y - v))
 end
 
-@inline function dsub2(x::T, y::Double{T}) where {T<:IEEEFloat}
+@inline function dsub2(x::vIEEEFloat, y::Double{<:vIEEEFloat})
     s = x - y.hi
     v = s - x
     Double(s, (x - (s - v)) + (-y.hi - v) - y.lo)
 end
 
-@inline function dsub2(x::Double{T}, y::T) where {T<:IEEEFloat}
+@inline function dsub2(x::Double{<:vIEEEFloat}, y::vIEEEFloat)
     s = x.hi - y
     v = s - x.hi
     Double(s, (x.hi - (s - v)) + (-y - v) + x.lo)
 end
 
-@inline function dsub2(x::Double{T}, y::Double{T}) where {T<:IEEEFloat}
+@inline function dsub2(x::Double{<:vIEEEFloat}, y::Double{<:vIEEEFloat})
     s = x.hi - y.hi
     v = s - x.hi
     Double(s, (x.hi - (s - v)) + (-y.hi - v) + x.lo - y.lo)
 end
 
-
+@inline (::Type{Vec{N,T}})(x::Vec{N,T}) where {N,T} = x
+@inline function SIMD.vifelse(b::Vec{N,Bool}, x::Double{T1}, y::Double{T2}) where {N,T<:Union{Float32,Float64},T1<:Union{T,Vec{N,T}},T2<:Union{T,Vec{N,T}}}
+    V = Vec{N,T}
+    Double(vifelse(b, V(x.hi), V(y.hi)), vifelse(b, V(x.lo), V(y.lo)))
+end
 
 if FMA_FAST
 
     # two-prod-fma
-    @inline function dmul(x::T, y::T) where {T<:IEEEFloat}
+    @inline function dmul(x::vIEEEFloat, y::vIEEEFloat)
         z = x * y
         Double(z, fma(x, y, -z))
     end
 
-    @inline function dmul(x::Double{T}, y::T) where {T<:IEEEFloat}
+    @inline function dmul(x::Double{<:vIEEEFloat}, y::vIEEEFloat)
         z = x.hi * y
         Double(z, fma(x.hi, y, -z) + x.lo * y)
     end
 
-    @inline dmul(x::T, y::Double{T}) where {T<:IEEEFloat} = dmul(y, x)
+    @inline dmul(x::vIEEEFloat, y::Double{<:vIEEEFloat}) = dmul(y, x)
 
-    @inline function dmul(x::Double{T}, y::Double{T}) where {T<:IEEEFloat}
+    @inline function dmul(x::Double{<:vIEEEFloat}, y::Double{<:vIEEEFloat})
         z = x.hi * y.hi
         Double(z, fma(x.hi, y.hi, -z) + x.hi * y.lo + x.lo * y.hi)
     end
 
     # x^2
-    @inline function dsqu(x::T) where {T<:IEEEFloat}
+    @inline function dsqu(x::T) where {T<:vIEEEFloat}
         z = x * x
         Double(z, fma(x, x, -z))
     end
 
-    @inline function dsqu(x::Double{T}) where {T<:IEEEFloat}
+    @inline function dsqu(x::Double{T}) where {T<:vIEEEFloat}
         z = x.hi * x.hi
         Double(z, fma(x.hi, x.hi, -z) + x.hi * (x.lo + x.lo))
     end
 
     # sqrt(x)
-    @inline function dsqrt(x::Double{T}) where {T<:IEEEFloat}
+    @inline function dsqrt(x::Double{T}) where {T<:vIEEEFloat}
         zhi = _sqrt(x.hi)
         Double(zhi, (x.lo + fma(-zhi, zhi, x.hi)) / (zhi + zhi))
     end
 
     # x/y
-    @inline function ddiv(x::Double{T}, y::Double{T}) where {T<:IEEEFloat}
+    @inline function ddiv(x::Double{<:vIEEEFloat}, y::Double{<:vIEEEFloat})
         invy = 1 / y.hi
         zhi = x.hi * invy
         Double(zhi, (fma(-zhi, y.hi, x.hi) + fma(-zhi, y.lo, x.lo)) * invy)
     end
 
-    @inline function ddiv(x::T, y::T) where {T<:IEEEFloat}
+    @inline function ddiv(x::vIEEEFloat, y::vIEEEFloat)
```

Comment:
So I'm guessing these changes in the type signature are required to operate on the vector version? 
I'm sure you are aware that the changes is not equivalent to the version on master.  
I just want to confirm.

Reply:
The old version forced `x` and `y` to be of the same type. The current version does not.
The reason for that change is that often one of `x` or `y` will be a scalar, and the other a `Vec`.
If you would like type checking to enforce that you don't mix `Float32` and `Float64`s, we could:
```julia
function foo(x::vIEEEFloat, y::vIEEEFloat)
    @assert eltype(x) == eltype(y)
    ...
```
or
```julia
function foo(x::T1, y::T2) where {T <: IEEEFloat, T1 <: Union{T,Vec{<:Any,T}}, T2 <: Union{T,Vec<:Any,T}}}
    ...
```
I haven't tested the second option, but I think something like it should work.